compiler/optimizer/VPHandlers.cpp

```diff
@@ -2019,77 +2028,133 @@ static TR::Node *findArrayIndexNode(OMR::ValuePropagation *vp, TR::Node *node, i
   bool usingAladd = (vp->comp()->target().is64Bit()
                      ) ?
           true : false;
-
-  if (usingAladd)
-     {
-     int32_t constValue;
-     if ((offset->getOpCode().isAdd() || offset->getOpCode().isSub()) &&
+   // Note: index node here refers to the array element index.
+   if (node->getFirstChild()->isDataAddrPointer())
+      {
+      TR::Node *offset2 = NULL;
+      if ((offset->getOpCode().isAdd() || offset->getOpCode().isSub()) &&
          offset->getSecondChild()->getOpCode().isLoadConst())
-        {
-        if (offset->getSecondChild()->getType().isInt64())
-           constValue = (int32_t) offset->getSecondChild()->getLongInt();
-        else
-           constValue = offset->getSecondChild()->getInt();
-        }
+         // aladd
+         //   aloadi (dataAddrPointer)
+         //       aload (array object)
+         //   ladd/lsub (offset)
+         //     lmul/lshl OR array_index (offset2)
+         //       i2l
+         //         iload array_index
+         //       lconst mul_stride/shift_amount
+         //     lconst some_constant
+         offset2 = offset->getFirstChild();
```

Comment:
I noticed that it ignores the addition, given it matches the `usingAladd`/non-offheap path it made sense to keep it consistent.

Reply:
It isn't consistent though. The off-heap path ignores any outer addition or subtraction, which can cause it to return a node that doesn't represent the right index. The existing code OTOH ignores an outer addition or subtraction only if it effectively adds exactly the array header size, which means that it will never miss a constant term in the index. The corresponding thing for the off-heap case would be to ignore an outer addition or subtraction of exactly zero, since the data address pointer already points to the first element

Regardless, given https://github.com/eclipse/omr/pull/7345#discussion_r1649334281, IMO this should be deleted rather than correctled anyway